linera-witty/src/memory_layout/layout.rs

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! Representation of the memory layout of complex types as a sequence of fundamental WIT types.
+
+use super::{element::LayoutElement, FlatLayout};
+use crate::{primitive_types::MaybeFlatType, util::Split};
+use frunk::{hlist::HList, HCons, HNil};
+
+/// Marker trait to prevent [`LayoutElement`] to be implemented for other types.
+pub trait Sealed {}
+
+/// Representation of the memory layout of complex types as a sequence of fundamental WIT types.
+pub trait Layout: Sealed + Default + HList {
+    /// The alignment boundary required for the layout.
+    const ALIGNMENT: u32;
+
+    /// Result of flattening this layout.
+    type Flat: FlatLayout;
+
+    /// Result of appending some `Other` layout to this layout.
+    type Append<Other: Layout>: Layout + Split<Self, Remainder = Other>;
+
+    /// Flattens this layout into a layout consisting of native WebAssembly types.
+    ///
+    /// The resulting flat layout does not have any empty items.
+    fn flatten(self) -> Self::Flat;
+
+    /// Appends some `other` layout with this layout, returning a new layout list.
+    fn append<Other>(self, other: Other) -> Self::Append<Other>
```

Comment:
This kinda makes sense at a high level but I'm also a bit surprised that we should need Generic Associated Types (GATs) here. Have you considered using good old combinators (like [futures](https://docs.rs/futures/latest/futures/future/fn.join.html)) instead of `frunk` + GATs + seals?

Is there a way to add tests that would demonstrate the intended use of the new definitions?

Reply:
Updated to remove the `Append` and `FlatAppend` generic associated types. They were replaced by using `std::ops::Add` instead (in later PRs), which `frunk` already supports.